src/adc.rs

```diff
@@ -0,0 +1,174 @@
+use crate::callback::{CallbackSubscription, SubscribableCallback};
+use crate::shared_memory::SharedMemory;
+use crate::syscalls_new;
+
+pub const DRIVER_NUM: usize = 0x0005;
+pub const BUFFER_SIZE: usize = 128;
+
+mod command {
+    pub const COUNT: usize = 0;
+    pub const START: usize = 1;
+    pub const START_REPEAT: usize = 2;
+    pub const START_REPEAT_BUFFER: usize = 3;
+    pub const START_REPEAT_BUFFER_ALT: usize = 4;
+    pub const STOP: usize = 5;
+}
+
+mod subscribe {
+    pub const SUBSCRIBE_CALLBACK: usize = 0;
+}
+
+mod allow {
+    pub const BUFFER: usize = 0;
+    pub const BUFFER_ALT: usize = 1;
+}
+
+#[derive(Debug)]
+pub enum Error {
+    NotSupported,
+    SubscriptionFailed,
+    Busy,
+    Invalid,
+    Fail,
+    Other(syscalls_new::Error),
+}
+
+impl From<syscalls_new::Error> for Error {
+    fn from(error: syscalls_new::Error) -> Error {
+        match error {
+            syscalls_new::Error::Nosupport => Error::NotSupported,
```

Comment:
Using this strategy, you only get the information about what kind of error has been thrown but you cannot tell which syscall caused the problem. I liked our old idea very much where each syscall type has its own return type, like `Result<usize, CommandError>`, `Result<CallbackSubscription, SubscriptionError>`, etc. and a common enum type, e.g. `enum TockError { Command(CommandError), Subscription(SubscriptionError), etc. }`